cpp/src/lance/arrow/dataset.cc

```diff
@@ -0,0 +1,198 @@
+const std::string kDataDir = "data";
+
+/// Get Manifest file path for a version.
+///
+/// \param base_uri the base uri of the dataset
+/// \param version optional version number. If not specified, returns the latest version.
+/// \return string path to the manifest file.
+std::string GetManifestPath(const std::string& base_uri, std::optional<uint64_t> version) {
+  if (version.has_value()) {
+    return fs::path(base_uri) / "_versions" / fmt::format("{}.manifest", version.value());
+  } else {
+    return fs::path(base_uri) / kLatestManifest;
+  }
+}
+
+class LanceDataset::Impl {
+ public:
+  Impl() = delete;
+
+  Impl(std::shared_ptr<::arrow::fs::FileSystem> filesystem,
+       std::string uri,
+       std::shared_ptr<lance::format::Manifest> m)
+      : fs(std::move(filesystem)), base_uri(std::move(uri)), manifest(std::move(m)) {}
+
+  std::string data_dir() const { return fs::path(base_uri) / kDataDir; }
+
+  std::shared_ptr<::arrow::fs::FileSystem> fs;
+  std::string base_uri;
+  std::shared_ptr<lance::format::Manifest> manifest;
+};
+
+LanceDataset::LanceDataset(std::unique_ptr<LanceDataset::Impl> impl)
+    : ::arrow::dataset::Dataset(impl->manifest->schema().ToArrow()), impl_(std::move(impl)) {}
+
+LanceDataset::LanceDataset(const LanceDataset& other)
+    : LanceDataset(std::make_unique<Impl>(*other.impl_)) {}
+
+LanceDataset::~LanceDataset() {}
+
+std::vector<std::shared_ptr<lance::format::DataFragment>> CreateFragments(
+    const std::vector<std::string>& paths, const lance::format::Schema& schema) {
+  std::vector<std::shared_ptr<lance::format::DataFragment>> fragments;
+  auto field_ids = schema.GetFieldIds();
+  for (auto path : paths) {
+    fragments.emplace_back(
+        std::make_shared<lance::format::DataFragment>(lance::format::DataFile(path, field_ids)));
+  }
+  return fragments;
+}
+
+::arrow::Status LanceDataset::Write(const ::arrow::dataset::FileSystemDatasetWriteOptions& options,
+                                    std::shared_ptr<::arrow::dataset::Scanner> scanner) {
+  const auto& base_dir = options.base_dir;
+  const auto data_dir = (fs::path(base_dir) / kDataDir).string();
+  auto& fs = options.filesystem;
+
+  // Load previous latest Manifest if any.
+  ARROW_ASSIGN_OR_RAISE(auto cur_dataset, LanceDataset::Make(options.filesystem, base_dir));
+
+  std::shared_ptr<lance::format::Manifest> manifest;
+  if (!cur_dataset) {
+    // This is a completely new dataset, create Manifest with version 1.
+    auto schema = std::make_shared<lance::format::Schema>(scanner->options()->dataset_schema);
+    manifest = std::make_shared<lance::format::Manifest>(schema);
+  } else {
+    // Bump the version
+    manifest = cur_dataset->impl_->manifest;
+    manifest->BumpVersion();
+  }
+  // Write manifest file
+  auto lance_option = options;
+  lance_option.base_dir = data_dir;
+  lance_option.existing_data_behavior = ::arrow::dataset::ExistingDataBehavior::kOverwriteOrIgnore;
```

Comment:
is this only for new dataset? or are we going to stop supporting the user supplied existing data behavior?

Reply:
this is just to re-use the FileSystemDataset::WRite function because we are writing multiple versions into the same directory. 

This behavior is not user-visible option.